python2-cython/challenge/basics.pyx

```diff
@@ -0,0 +1,252 @@
+"""Definitions of Python extension types for basics module
+"""
+
+from libcpp cimport bool
+from libcpp.memory cimport shared_ptr
+from cython.operator cimport dereference as deref
+from cpython.object cimport Py_EQ, Py_NE
+
+from _basics cimport move
+from _basics cimport compare as _compare
+from _basics cimport adjacent as _adjacent
+from basics cimport Doodad
+from _basics cimport Doodad as _Doodad
+from _basics cimport Secret as _Secret
+from _basics cimport WhatsIt as _WhatsIt
+
+cpdef compare(Secret a, Secret b):
+    return _compare(deref(a.thisptr), deref(b.thisptr))
+
+cpdef adjacent(Secret a, Secret b):
+    return _adjacent(deref(a.thisptr), deref(b.thisptr))
+
+cdef _WhatsIt from_tuple(tuple t):
+    """Helper function only visible from Cython.
+    """
+
+    cdef _WhatsIt w = _WhatsIt(t[0], t[1])
+
+    return w
+
+cdef class Secret:
+    """Opaque type only for use by Doodad.
+    """
+
+    cdef const _Secret *thisptr
+
+cdef class Doodad:
+    """Python interface to C++ type Doodad.
+
+    Parameters
+    ----------
+    name : str | tuple
+        When tuple, a pair of name, value. Otherwise a string.
+    value : int
+        A value
+    """
+
+    def __init__(self, name=None, value=1, init=True):
+        if init:
+            if isinstance(name, tuple):
+                self.thisptr.reset(new _Doodad(from_tuple(name)))
+            elif isinstance(name, str):
+                self.thisptr.reset(new _Doodad(name, value))
+            else:
+                raise TypeError("a string or tuple is required")
```

Comment:
I'd love to. But I haven't found a way how to yet. The problem is that this is one of the cases where Cython is different from Python. Cython had `__new__` before Python came up with it and it meant something different. It performed pre-initialization at the C level (e.g. malloc). Now that Python has `__new__` this has been deprecated and renamed as `__cinit__`. But Cython doesn't have the real Python `__new__`. Or at least it doesn't look like it. When I use `d = Doodad.__new__(Doodad)` to skip `__init__` instead of the ugly `d = Doodad(init=False)` it doesn't know about it's C types and thus the subsequent assignment of the `shared_ptr` tries to cast to a Python object which it can't so the cythonizing fails there.